lib/kubernetes-deploy/render_task.rb

```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+require 'tempfile'
+
+require 'kubernetes-deploy/renderer'
+require 'kubernetes-deploy/template_discovery'
+
+module KubernetesDeploy
+  class RenderTask
+    def initialize(logger:, current_sha:, template_dir:, bindings:)
+      @logger = logger
+      @template_dir = template_dir
+      @renderer = KubernetesDeploy::Renderer.new(
+        current_sha: current_sha,
+        bindings: bindings,
+        template_dir: @template_dir,
+        logger: @logger,
+      )
+    end
+
+    def run(*args)
+      run!(*args)
+      true
+    rescue KubernetesDeploy::FatalDeploymentError
+      false
+    end
+
+    def run!(stream, only_filenames = [])
+      @logger.phase_heading("Initializing render task")
+
+      filenames = if only_filenames.empty?
+        TemplateDiscovery.new(@template_dir).templates
+      else
+        only_filenames
+      end
+
+      @logger.phase_heading("Rendering template(s)")
+      validate_configuration(filenames)
+
+      filenames.each { |filename| render_filename(filename, stream) }
+
+      @logger.summary.add_action("Successfully rendered #{filenames.size} template(s)")
+      @logger.print_summary(:success)
+    rescue KubernetesDeploy::FatalDeploymentError
+      @logger.print_summary(:failure)
+      raise
+    end
+
+    private
+
+    def render_filename(filename, stream)
+      file_content = File.read(File.join(@template_dir, filename))
+      rendered_content = @renderer.render_template(filename, file_content)
+      YAML.load_stream(rendered_content) do |doc|
+        stream.puts YAML.dump(doc)
+      end
+      @logger.info("Rendered #{File.basename(filename)}")
+
+    rescue KubernetesDeploy::InvalidTemplateError => exception
+      log_invalid_template(filename, exception)
+      raise
+    rescue Psych::SyntaxError => exception
+      log_invalid_template(filename, exception)
+      raise InvalidTemplateError.new("Template is not valid YAML", filename: filename)
+    end
+
+    def validate_configuration(filenames)
+      @logger.info("Validating configuration")
+      errors = []
+
+      if filenames.empty?
+        errors << "no templates found in template dir #{@template_dir}"
+      end
+
+      absolute_template_dir = File.expand_path(@template_dir)
+
+      filenames.each do |filename|
+        absolute_file = File.expand_path(File.join(@template_dir, filename))
+        if !File.exist?(absolute_file)
+          errors << "Filename \"#{absolute_file}\" could not be found"
+        elsif !File.file?(absolute_file)
+          errors << "Filename \"#{absolute_file}\" is not a file"
+        elsif !absolute_file.start_with?(absolute_template_dir)
+          errors << "Filename \"#{absolute_file}\" is outside the template directory"
```

Comment:
I did that on purpose because I think that the interaction between `--template-dir` and templates passed as ARGV is not altogether clear to the uninitiated. This tool doesn't allow subfolders other than partials I think, so it's not like you can actually pass any old relative path as an ARGV, especially something with like `../` in it, so I think it's worth being extra clear. For bonus points you can copy paste that absolute path directly and `cat` it or whatever to understand exactly what the tool is doing.